xchange-gdax/src/main/java/org/knowm/xchange/gdax/service/GDAXMarketDataServiceRaw.java

```diff
@@ -90,7 +97,74 @@ public GDAXProductBook getGDAXProductOrderBook(CurrencyPair currencyPair, int le
       throw handleError(e);
     }
   }
+  
+  static Instant lastCall = null;
+  static boolean rateLimited = false;
+  
+  public GDAXTrades getGDAXTradesExtended(CurrencyPair currencyPair, Long after, Integer limit) throws IOException {	
+	for (;;) {
+	    try {
+	      if (rateLimited) {
+	    	  long delta = Duration.between(lastCall, Instant.now()).toMillis();
+	    	  log.debug("Last call {} ms ago", delta);
+	    	  
+	    	  if (delta < 333) {
+	    		  try {
+	    			log.debug("Sleeping for {}ms", 333-delta);
+					Thread.sleep(333-delta);
+					
+				  } catch (InterruptedException e) {
+					// TODO Auto-generated catch block
+					e.printStackTrace();
+				  }
+	    	  } else if (delta > 3000) {
+	    		  log.debug("Clearing rate limiter");
+	    		  rateLimited = false;
+	    	  }
+	      }
+	    	
+	      lastCall = Instant.now();
+	      
+	      GDAXTrades gdaxTrades = this.gdax.getTradesPageable(
+	          currencyPair.base.getCurrencyCode(), 
+	          currencyPair.counter.getCurrencyCode(),
+	          after,
+	          limit);
+	      
+	      
+	      log.debug("GDAXTrades: earliest={}, latest={}, {}", 
+	    		  gdaxTrades.getEarliestTradeId(), 
+	    		  gdaxTrades.getLatestTradeId(), 
+	    		  gdaxTrades);
+	      
+	      return gdaxTrades;
+	    	
+	    } catch (GDAXException e) {
+	    	if (e.getHttpStatusCode() == 429) {
+		    	log.debug("Rate limit exceeded, sleeping for 1000ms");
+		    	
+		    	rateLimited = true;
+		    	
+		    	try {
+					Thread.sleep(1000);
```

Comment:
this should not be done at this level, remove the sleep calls and do call rate limiting in the client code

Reply:
Afraid I disagree with this.  Rate limiting is a lower level function, ideally lower than this, but until we can hook it into the rescu package this is the best we can do.  Client code should NOT have to be concerned with the gdax rate limits I think.

Reply:
All that is certainly true. But, rather surprisingly, this sketchy solution actually works in testing rather well.

I'm not aware of any other method that specifically needs to fire large numbers of calls at gdax.

Reply:
I'm not claiming it does not work, if it didn't I would not call it a 'solution' :o)
getGDAXProductOrderBook, getGDAXTrades, getGDAXHistoricalCandles
and similar methods in every other module 
we need a global solution, not case by case boiler code

Reply:
Adam,  if I change this then the client has to be responsible for not only all retries but also chaining the results of all the different calls together.  This really does seem an unnecessary load to place on the client code.

I appreciate your point of view about needing a global solution, but until we get one then case-by-case boilerplate is the best we can do.

I really am very reluctant to remove this.

Reply:
I agree that client code should be responsible to retries. We've had conversations about this in other threads. Mainly, the scope of XChange is to provide gateways to all the REST endpoints. One sorta global solution that clients can use is the `Retries` class, but I haven't personally used it yet. Can you try it out?